core/src/avm2/globals/toplevel.rs

```diff
@@ -141,30 +141,225 @@ pub fn parse_int<'gc>(
     }
 }
 
+/// Strips leading whitespace characters.
+fn skip_spaces(s: &mut &WStr) {
+    *s = s.trim_start_matches(|c| {
+        matches!(
+            c,
+            0x20 | 0x09 | 0x0d | 0x0a | 0x0c | 0x0b | 0x2000
+                ..=0x200b | 0x2028 | 0x2029 | 0x205f | 0x3000
+        )
+    });
+}
+
+/// Consumes an optional sign character.
+/// Returns whether a minus sign was consumed.
+fn parse_sign(s: &mut &WStr) -> bool {
+    if let Some(after_sign) = s.strip_prefix(b'-') {
+        *s = after_sign;
+        true
+    } else if let Some(after_sign) = s.strip_prefix(b'+') {
+        *s = after_sign;
+        false
+    } else {
+        false
+    }
+}
+
+/// Converts a `WStr` to an `f64`.
+///
+/// This function might fail for some invalid inputs, by returning `None`.
+///
+/// `strict` typically tells whether to behave like `Number()` or `parseFloat()`:
+/// * `strict == true` fails on trailing garbage, but interprets blank strings (which are empty or consist only of whitespace characters) as zero.
+/// * `strict == false` ignores trailing garbage, but fails on blank strings.
+fn string_to_f64(mut s: &WStr, swf_version: u8, strict: bool) -> Option<f64> {
+    // Allow leading whitespace characters.
+    skip_spaces(&mut s);
+
+    if s.is_empty() {
+        // A blank string. Handle it as described above.
+        return if strict { Some(0.0) } else { None };
+    }
+
+    let is_negative = parse_sign(&mut s);
+    let after_sign = s;
+
+    // Count digits before decimal point.
+    s = s.trim_start_matches(|c| (c as u8).is_ascii_digit());
+    let mut total_digits = after_sign.len() - s.len();
+
+    // Stop parsing at null character.
+    if s.starts_with(b'\0') {
```

Comment:
Is there a reason why we can't truncate the string right at the start of the function instead of checking at specific points throughout?

Reply:
That what `avmplus` did. But after thinking about it, I came into a conclusion that these stops are only relevant for the `strict` mode check below (which isn't used yet - should be used by `Number(...)`). So I updated this PR by replacing the null character stops with a `strict` mode check that permits a null character. Once `Number(...)` will use `string_to_f64` in a future PR, `Number("10" + String.fromCharCode(0))` will be considered valid.